src/osp/Active/SysNewton.cpp

```diff
@@ -525,128 +548,45 @@ Vector4 SysNewton::compute_body_CoM(ActiveScene& rScene,
-            inertiaTensor[2][2] = principalAxes.z();
-
-            /* CAUTION: this computation is done with respect to the part's origin.
-                        If the part CoM does not coincide with the origin, unexpected
-                        behavior may result. Either add a loop to compute the part's
-                        local center of mass above, or require parts to set their
-                        origin to be the CoM.
-            */
-            Matrix3 rotation = childMatrix.rotation();
-            Vector3 offset = childMatrix.translation();
-            I += phys::transform_inertia_tensor(inertiaTensor, mass, offset, rotation);
-        }
+        Matrix3 rotation = childTransform.rotation();
+        Vector3 offset = (childTransform.translation() + childCoM.xyz()) - centerOfMass.xyz();
+        I += phys::transform_inertia_tensor(
+            childInertia, childCoM.w(), offset, rotation);
 
-        nextChild = childHier.m_siblingNext;
+        nextChild = rScene.reg_get<ACompHierarchy>(nextChild).m_siblingNext;
     }
 
-    return I;
-}
-
-/*
- NOTE: The problem here is that the inertia tensor calculation depends on mass
-       distribution, but our point-mass-plus-colliders model doesn't really allow
-       for the specification of mass distribution. For now, the solution is for
-       compute_mass_inertia() to find the volume of the part's colliders and
-       compute the average density of the part, which is then used to assume a
-       uniform mass distribution. Future solutions may involve requiring
-       colliders to specify density.
-
- TODO: This function is not recursive yet, because rigid bodies are currently
-       only one level deep. If that changes, this function will need to be
-       updated accordingly.
-*/
-Matrix3 SysNewton::compute_body_inertia(ActiveScene& rScene, ActiveEnt root,
-    Vector3 centerOfMass, Matrix4 currentTransform)
-{
-    Matrix3 localI{0.0f};
-    float localMass = 0.0f;
-
-    for (ActiveEnt nextChild = root; nextChild != entt::null;)
+    // Include entity's own inertia, if it has a mass and volume from which to compute it
+    auto const* mass = rScene.reg_try_get<ACompMass>(entity);
+    auto const* shape = rScene.reg_try_get<ACompShape>(entity);
+    if (mass != nullptr && shape != nullptr)
     {
-        auto const &childHier = rScene.reg_get<ACompHierarchy>(nextChild);
-        auto const &childTransform = rScene.reg_get<ACompTransform>(nextChild);
+        Matrix4& transform = rScene.reg_get<ACompTransform>(entity).m_transform;
+        Vector3 principalAxes = phys::collider_inertia_tensor(
+            shape->m_shape, transform.scaling(), mass->m_mass);
 
-        Matrix4 childMatrix = currentTransform * childTransform.m_transform;
+        Matrix3 localInertiaTensor{};
+        localInertiaTensor[0][0] = principalAxes.x();
+        localInertiaTensor[1][1] = principalAxes.y();
+        localInertiaTensor[2][2] = principalAxes.z();
```

Comment:
I'm slightly surprised that the Matrix3 class doesn't have a static function to do this for you.

Reply:
This seems like a common enough usage to warrant creating a function for it.

Reply:
Too bad we can't inject a constructor into it or something, glm has this functionality (to initialize a diagonal matrix from a vector)